File: src/chipschallenge/blockreactions/ComputerChipTo.java

package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Block.Type;
import chipschallenge.BlockContainerFullException;


public class ComputerChipTo extends BlockReaction {

    private ComputerChipTo() {}
    private static ComputerChipTo mInstance = null;

    public static synchronized ComputerChipTo getInstance() {
        if (mInstance == null) {
            mInstance = new ComputerChipTo();
        }
        return mInstance;
    }

    @Override
    public void react(Block moving, Block standing) throws BlockContainerFullException {
        game().takeChip();
        standing.replace(createBlock(Type.FLOOR));
    }

    @Override
    public boolean canMove(Block moving, Block standing) {
        return moving.isChip();
    }

}

File: ChipsChallenge.Shared/Blockreactions/ComputerChipTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    public class ComputerChipTo : NoSlipReaction
    {
        private static ComputerChipTo instance;

        private ComputerChipTo()
        {
        }

        public static ComputerChipTo Instance
        {
            get
            {
                lock (typeof(ComputerChipTo))
                {
                    return instance ?? (instance = new ComputerChipTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            if (moving.Chip)
            {
                Game().TakeChip();
                standing.Replace(CreateBlock(Block.Type.FLOOR));
                Sound().Play(Shared.Sound.TakeChip);
            }
        }

        public override bool canMove(Block moving, Block standing)
        {
            return moving.Chip;
        }
    }
}